src/Agda2Hs/Compile/RuntimeCheckUtils.hs

```diff
@@ -0,0 +1,252 @@
+-- Takes telescope of type to check.
+checkRtc' :: NameIndices -> Telescope -> C (NameIndices, RtcResult')
+checkRtc' idcs tel = do
+  ourChks <- mapM (uncurry createGuardExp) topLevelErased
+  (belowIdcs, belowChks) <- mapAccumLM checkRtc'' idcs call
+  (belowIdcs,)
+    <$> if all isJust belowChks && all isJust ourChks
+      then
+        let (theirLhs, theirRhs, decls) = unzip3 $ catMaybes belowChks
+            theirDecls = concat decls
+            theirChks = catMaybes ourChks
+         in if null theirDecls && null ourChks
+              then return NoneErased'
+              else return Checkable' {..}
+      else return Uncheckable'
+  where
+    doms = telToList tel
+    telsUpTo = map (\i -> fst $ splitTelescopeAt i tel) [0 ..]
+    -- Partition out arguments that are erased and at top level (those we will attempt to check)
+    (topLevelErased, call) = partition (checkTopLevelErased . fst) $ zip doms telsUpTo
+
+-- Check a single type for runtime checkability.
+-- Accumulates on name indices for `go` function and `a` argument.
+-- Takes domain of type and telescope up to that point for context.
+-- If checkable, returns lhs and rhs at that point plus declarations from checks below.
+checkRtc'' :: NameIndices -> (Dom (ArgName, Type), Telescope) -> C (NameIndices, Maybe (Hs.Pat (), Hs.Exp (), [Hs.Decl ()]))
+checkRtc'' (goIdx, argIdx) (d, tUpTo) = do
+  ((theirGoIdx, theirArgIdx), chk) <- addContext tUpTo (checkRtc' (goIdx, ourArgIdx) tAt)
+  case chk of
+    NoneErased' -> return ((theirGoIdx, theirArgIdx), Just (ourLhs, hsVar arg, []))
+    Uncheckable' -> return ((theirGoIdx, theirArgIdx), Nothing)
+    Checkable' {..} -> do
+      let go = "go" ++ show theirGoIdx
+          conflicts = tAtNames `intersect` [go, arg, up]
+      unless (null conflicts) $ errorNameConflict $ head conflicts
+      let rhs = hsVar up `eApp` theirRhs
+          maybeRtc = createRtc (hsName go) (hsPat up : theirLhs) theirChks rhs $ binds theirDecls
+          (ourGoIdx, ourRhs, ourRtc) =
+            if null theirChks
+              then -- inline if nothing to check at this level (consumes no `goIdx`)
+                (theirGoIdx, Hs.Lambda () theirLhs $ hsVar arg `eApp` theirRhs, theirDecls)
+              else (succ theirGoIdx, hsVar go `eApp` [hsVar arg], [maybeRtc])
+      return ((ourGoIdx, theirArgIdx), Just (ourLhs, ourRhs, ourRtc))
+  where
+    tAt = telify d
+    tAtNames = map (fst . unDom) $ telToList tAt
+    name = fst $ unDom d
+    -- Use arg name if available, otherwise insert one (consumes one on `argIdx`)
+    (arg, ourArgIdx) = if name == "_" then ("a" ++ show argIdx, succ argIdx) else (name, argIdx)
+    ourLhs = hsPat arg
+    up = "up"
+
+---- Small single-mention convenience functions
+
+-- Gather telescope from domain
+telify :: Dom (a, Type) -> Telescope
+telify = theTel . telView' . snd . unDom
+
+-- Check a type is erased and at top level; in this case, it should be checked.
+checkTopLevelErased :: Dom (a, Type) -> Bool
```

Comment:
This is somewhat problematic: since https://github.com/agda/agda2hs/pull/304/ `agda2hs` no longer purely relies on erasure annotations to determine whether to drop an argument.
Ideally, for consistency with the rest of agda2hs we would use the result of `compileDom` to determine whether or not an argument is erased. Otherwise, there is still a risk we miss preconditions in `Prop`, for example